### PR TITLE
Web only version (gRPC related examples removed)

### PR DESCRIPTION
**Pull request by tgrospic**
Also added [htm](https://github.com/developit/htm) library to write render functions with template literals which looks like HTML.

### CODE REVIEW

File: src/web/index.html [hunk 1]
```diff
@@ -23,6 +23,11 @@
   "> -->
 </head>
 <body>
+  <div class="dev-info-box">
+    This is testing version of the page. For official version please visit
```
Comment (dckc) [inline src/web/index.html:27]: official? What's official about the other one? you might call it "production"

File: src/lib.js [hunk 2]
```diff
@@ -9,5 +22,29 @@ export const decodeBase16 = hexStr => {
   return Uint8Array.from(resArr)
 }
 
+/**
+ * Encode base 16 string to base 58.
+ *
+ * @param {string} hexStr
+ * @returns {string}
+ */
+export const encodeBase58 = hexStr => {
```
Comment (dckc) [inline src/lib.js:31]: A name like `lib.js` or `utils.js` is a sort of code smell.

I have been using [hex.js](https://github.com/rchain-community/liquid-democracy/blob/dev-tools/rclient/hex.js) that exports `Basic16` with `encode` and `decode` static methods. (I think that particular copy only has encode, but you get the idea).

I suppose `Base58.encode` and `.decode` could belong in the same file if it were called `codecs.js`.

File: src/rev-address.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
+// @ts-check
+import { keccak256 } from 'js-sha3'
+import blake from 'blakejs'
+import { ec } from 'elliptic'
+import { decodeBase16, encodeBase58, encodeBase16, decodeBase58safe } from './lib'
+
+const secp256k1 = new ec('secp256k1')
```
Comment (dckc) [inline src/rev-address.js:7]: Is there any mutable state inside this `secp256k1`?

File: src/rev-address.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+// Algorithm to generate ETH and REV address is taken from RNode source
+// https://github.com/rchain/rchain/blob/bf7a30e1/rholang/src/main/scala/coop/rchain/rholang/interpreter/util/AddressTools.scala#L47
+
+// Prefix as defined in https://github.com/rchain/rchain/blob/c6721a6/rholang/src/main/scala/coop/rchain/rholang/interpreter/util/RevAddress.scala#L13
+const prefix = { coinId : "000000", version: "00" }
+
+/**
+ * @typedef {Object} RevAccount - Represents different formats of REV address
```
Comment (dckc) [inline src/rev-address.js:16]: You can also write this whole typedef in one line:
```js
 * @typedef { {privKey?: string, pubKey?: string, ethAddr?: string, revAddr: string } } RevAccount
```

File: src/rev-address.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+/**
+ * Get REV address from ETH address.
+ *
+ * @param {string} ethAddrRaw
+ * @returns {string}
+ */
+export const getAddrFromEth = ethAddrRaw => {
+  const ethAddr = ethAddrRaw.replace(/^0x/, '')
+  if (!ethAddr || ethAddr.length !== 40) return
```
Comment (dckc) [inline src/rev-address.js:31]: does this typecheck with `@returns { string }`?

File: src/rev-address.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+  return { pubKey, ...addr }
+}
+
+/**
+ * Verify REV address
+ * @param {string} revAddr
+ */
+export const verifyRevAddr = revAddr => {
+  const revBytes = decodeBase58safe(revAddr)
+  if (!revBytes) return
```
Comment (dckc) [inline src/rev-address.js:93]: again, does this typecheck? it should throw, no? Or change the return type to `string | undefined`.

File: src/rev-address.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+/**
+ * Generates new private and public key, ETH and REV address.
+ *
+ * @returns {RevAccount}
+ */
+export const newRevAccount = () => {
+  // Generate new key and REV address from it
+  const key     = secp256k1.genKeyPair()
```
Comment (dckc) [inline src/rev-address.js:113]: OUCH! ambient access to a random number generator!

File: src/rev-address.js (hunk 3 above)
Comment (dckc) [reply src/rev-address.js:7]: yes, it seems to include a random number generator.

File: src/web/controls/address-ctrl.js [hunk 8]
```diff
@@ -66,73 +50,76 @@ export const addressCtrl = (st, {wallet, onAddAccount}) => {
...
+    <div class="ctrl address-ctrl">
+      <h2>REV wallet (import REV address, ETH address, public/private key, Metamask)</h2>
+      ${description}
+
+      <!-- Input textbox -->
+      <div ...${labelStyle(text)}>${labelSource}</div>
+      <input type=text autocomplete=nono placeholder=${labelSource} value=${text} oninput=${addrKeyPressEv} />
+
+      <!-- New accounts -->
+      ${ethDetected && html`
```
Comment (dckc) [inline src/web/controls/address-ctrl.js:79]: OCAP-OUCH! global mutable state

**Comment by dckc** [comment]
The htm stuff in 8d0aa4fa is great!

File: src/rev-address.js (hunk 7 above)
Comment (dckc) [reply src/rev-address.js:113]: I raised a separate issue: https://github.com/tgrospic/rnode-client-js/issues/18

File: src/rnode-sign.js [hunk 9]
```diff
@@ -1,13 +1,45 @@
+// @ts-check
 import blake from 'blakejs'
 import { ec } from 'elliptic'
 import jspb from 'google-protobuf'
 
+/**
+ * These deploy types are based on protobuf specification which must be
+ * used to create the hash and signature of deploy data.
+ * Deploy object sent to Web API is slightly different, see [rnode-web.js](rnode-web.js).
```
Comment (dckc) [inline src/rnode-sign.js:9]: very nice. I spent a lot of time puzzling over these types, looking at scala code, etc.

See also https://github.com/rchain-community/liquid-democracy/blob/dev-tools/rclient/types.js WIP

File: src/rnode-web.js [hunk 10]
```diff
@@ -37,15 +75,28 @@ const makeRNodeHttpInternal = domFetch => async (httpUrl, apiMethod, data) => {
   // Add status if server error
   if (!resp.ok) {
     const ex = Error(result)
+    // @ts-ignore
```
Comment (dckc) [inline src/rnode-web.js:78]: yeah, I did the same thing. hm.

File: src/web/index.js [hunk 11]
```diff
@@ -12,7 +13,7 @@ const { fetch } = window
 const log = pageLog({log: logOrig, document})
 
 // Make RNode web API client / wrapped around DOM fetch
-const rnodeWeb = makeRNodeWeb({fetch})
+const rnodeWeb = makeRNodeWeb({fetch, now: Date.now})
```
Comment (dckc) [inline src/web/index.js:16]: yay!

**Review by dckc — APPROVED** [review]
excellent.

A few OCap no-nos remain, but those are orthogonal to these changes, so I'm raising separate issues.

File: src/lib.js (hunk 2 above)
Comment (dckc) [reply src/lib.js:31]: see https://github.com/rchain-community/rchain-api/blob/master/src/codec.js

**Comment by dckc** [comment]
I hope this lands soon. What's left to do?

**Comment by tgrospic** [comment]
Subsumed by newer PR #21 with full TypeScript support.

**Pull request closed without merging**